pkg/controller/types.go

```diff
@@ -0,0 +1,118 @@
+package controller
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type SynchronizationState uint8
+
+const (
+	SynchronizationStateUnknown SynchronizationState = iota
+	SynchronizationStateSuccessful
+	SynchronizationStateFailed
+)
+
+func (s *SynchronizationState) String() string {
+	switch *s {
+	case SynchronizationStateUnknown:
+		return "Unknown"
+	case SynchronizationStateSuccessful:
+		return "Successful"
+	case SynchronizationStateFailed:
+		return "Failed"
+	}
+	panic(fmt.Sprintf("Invalid SynchronizationState: %d", s))
+}
+
+func (s *SynchronizationState) UnmarshalText(text []byte) error {
+	str := string(text)
+	switch str {
+	case "Successful":
+		*s = SynchronizationStateSuccessful
+	case "Failed":
+		*s = SynchronizationStateFailed
+	case "Unknown":
+		*s = SynchronizationStateUnknown
+	default:
+		return fmt.Errorf("invalid SynchronizationState: %v", str)
+	}
+	return nil
+}
+
+func (s *SynchronizationState) MarshalText() (text []byte, err error) {
+	return []byte(s.String()), nil
+}
+
+// +kubebuilder:object:generate=true
+
+type NaisStatus struct {
+	// SynchronizationHash is the hash of the object most recently successfully synchronized
+	SynchronizationHash string `json:"synchronizationHash,omitempty"`
+	// SynchronizationTime is the last time the Status subresource was updated
+	SynchronizationTime *metav1.Time `json:"synchronizationTime,omitempty"`
+	// SynchronizationState is the result of the most recent synchronization
+	SynchronizationState SynchronizationState `json:"synchronizationState,omitempty"`
+	// SynchronizedGeneration is the generation most recently successfully synchronized
+	SynchronizedGeneration int64 `json:"synchronizedGeneration,omitempty"`
+	// ObservedGeneration is the generation most recently observed
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+	// Represents the latest available observations of the resource current state.
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+}
+
+type ReconcileResult struct {
+	// Requeue indicates whether the reconciliation should be requeued for a new attempt in a few seconds
+	Requeue bool
+
+	// Conditions should be a list of possible failure conditions
+	//
+	// Status == ConditionTrue means a failure has happened
+	Conditions []metav1.Condition
+
+	// State should indicate the overall state of the reconciliation
+	//
+	// If State == SynchronizationStateFailed, Conditions should contain at least one Condition with Status == ConditionTrue
+	State SynchronizationState
+}
+
+type NaisResource interface {
+	client.Object
+
+	// ApplyDefaults must apply default values to the resource
+	ApplyDefaults() error
+
+	// Hash must return a hash of the relevant parts of the resource
+	Hash() (string, error)
+
+	// GetStatus must return a non-nil pointer to the resource status attached to this resource
+	GetStatus() *NaisStatus
```

Comment:
Is returning a concrete type the proper solution? This limits how the consumer can utilize status for custom fields etc

Most of the fields defined above could probably be `metav1.Condition`

Reply:
Should it instead be an interface? How would that interface look :thinking: 

I'll play around with it to see what I can work out.

Some (all)  of the fields on `NaisStatus` that aren't `metav1.Condition` are there for familiarity. Some of our users expect to find some of these fields on the resource when they are troubleshooting, so I elected to keep them, even though the information would in some cases be better expressed in Conditions.

My understanding (but I'll go back and check) is that this doesn't limit the use of custom fields for the reconciler, they would still be defined and saved to the cluster. See https://github.com/nais/liberator/pull/190/files#diff-def652c66af56bf303fed68167f92e72e9154d991f142b60bfef067929b96a44R140 for how it looks on `TopicStatus`.

Reply:
I have missed the way the status get's updated, so I think you're correct :)